packages/component/src/Attachment/Text/private/MarkdownTextContent.tsx

```diff
@@ -234,21 +235,24 @@ const MarkdownTextContent = memo(({ activity, children, markdown }: Props) => {
         </LinkDefinitions>
       )}
       <div className="webchat__text-content__activity-actions">
-        {activity.type === 'message' && isBasedOnSoftwareSourceCode(messageThing) && messageThing.isBasedOn.text ? (
-          <ActivityViewCodeButton
-            className="webchat__text-content__activity-view-code-button"
-            code={messageThing.isBasedOn.text}
-            isAIGenerated={isAIGeneratedActivity(activity)}
-            language={messageThing.isBasedOn.programmingLanguage}
-            title={messageThing.isBasedOn.programmingLanguage}
-          />
-        ) : null}
-        {activity.type === 'message' && activity.text && messageThing?.keywords?.includes('AllowCopy') ? (
-          <ActivityCopyButton className="webchat__text-content__activity-copy-button" targetRef={contentRef} />
-        ) : null}
-        {activity.type === 'message' && feedbackActionsPlacement === 'activity-actions' && (
-          <ActivityFeedback activity={activity} />
-        )}
+        {/* TODO: [P2] Fold the ActivityActionsDecorator into the ActivityDecorator to unlock more flexibility */}
+        <ActivityActionsDecorator activity={activity}>
+          {activity.type === 'message' && isBasedOnSoftwareSourceCode(messageThing) && messageThing.isBasedOn.text ? (
```

Comment:
It feels like each of button in this 3 button set should be their own middleware.

Say...

```ts
() => next => request => {
  if (activity.type === 'message' && isBasedOnSoftwareSourceCode(messageThing) && messageThing.isBasedOn.text) {
    return <><ActivityViewCodeButton ... />{next(request)}</>;
  }

  return next(request);
}
```

Reply:
Except this is v2-style middleware, so it should return a component. We also don't want to expose activity on the request level, do we? Or, are you speaking about a dedicated middleware stack for activity actions?

Reply:
Dedicated middleware stack for activity actions.